opengovsg/postmangovsg
# Tesseract backend route

## Problem

A new endpoint was needed to handle download and process files from vault urls.

Closes #1001

## Solution

**Features**:

- Create `/tesseract` route for email, sms and telegram
- Endpoint accepts a vault url, url is validated using `VAULT_DOMAIN_NAME` and checks for existence of expiry and datasetname
- Stream file from s3 presigned url and process file similar to `/upload/complete` middleware
- Vault file metadata is saved in the 's3Object' column, with bucket as `VAULT_BUCKET_NAME` env var
- Include bucket in upload status endpoint response - to be used to identify source of recipient file (user uploaded or vault)

## Tests

- Generate a s3 pre-signed url using the aws cli for an object in `file-staging.postman.gov.sg`
- Use Postman api to call `/campaign/{campaignId}/tesseract` with the url in the body for sms, email and telegram channel
- Observer request response and check that db contains processed file records 
- Repeat for invalid urls and malformed recipient files

**Expired URL**
https://storage-test.vault.gov.sg/daniel%40open.gov.sg-req_379492e7-916e-42e9-9d98-842b316e0ef6-data_c8277f7e-52f9-4ddd-abec-3c6aebb90179?AWSAccessKeyId=AKIATNLXS5MLIQORMRW4&response-content-disposition=attachment%3B%20filename%3D%22testpostmandataset.csv%22&Expires=1615202392&Signature=3iD7pHVcGRWuTHd%2B3lszIvHKwMA%3D&datasetname=testpostmandataset
**Dataset with column Recipient**
https://storage-test.vault.gov.sg/lennard%40open.gov.sg-req_379492e7-916e-42e9-9d98-842b316e0ef6-data_c8277f7e-52f9-4ddd-abec-3c6aebb90179?AWSAccessKeyId=AKIATNLXS5MLIQORMRW4&Expires=1616048834&Signature=7Zmstjfoi7025YIt98QxeSB5rUY%3D&response-content-disposition=attachment%3B%20filename%3D%22testpostmandataset.csv%22&datasetname
**Dataset without column Recipient**
https://storage-test.vault.gov.sg/norecipient%40open.gov.sg-req_379492e7-916e-42e9-9d98-842b316e0ef6-data_c8277f7e-52f9-4ddd-abec-3c6aebb90179?AWSAccessKeyId=AKIATNLXS5MLIQORMRW4&Expires=1616048923&Signature=mije5RUV3K%2BXD2AHXURyaT8c6rE%3D&response-content-disposition=attachment%3B%20filename%3D%22testpostmandataset.csv%22&datasetname=testpostmandataset

## Deploy Notes

**New environment variables**:

- `VAULT_BUCKET_NAME ` : Bucket name attribute to be stored for vault datasets
- `VAULT_DOMAIN_NAME` : Vault domain name expected for vault urls
